google_guest_agent/command/command_monitor.go

```diff
@@ -0,0 +1,263 @@
+	c.signal <- "STOP"
+}
+
+// Start signals the server to start listening for commands.
+func (c *Server) Start() {
+	c.signal <- "START"
+}
+
+func (c *Server) listen(ctx context.Context) error {
+	// Do not call this function without holding c.srvMu. Returns when c.srv is
+	// set and the listener is accepting connections in another goroutine
+	srv, err := listen(ctx, c.pipe, c.pipeMode, c.pipeGroup)
+	if err != nil {
+		return err
+	}
+	go func() {
+		defer srv.Close()
+		for {
+			if ctx.Err() != nil {
+				return
+			}
+			conn, err := srv.Accept()
+			if err != nil {
+				if err == net.ErrClosed {
+					break
+				}
+				logger.Infof("error on connection to pipe %s: %v", c.pipe, err)
+				continue
+			}
+			go func(conn net.Conn) {
+				defer conn.Close()
+				// Go has lots of helpers to do this for us but none of them return the byte
+				// slice afterwards, and we need it for the handler
+				var b []byte
+				r := bufio.NewReader(conn)
+				var depth int
+				deadline := time.Now().Add(c.timeout)
+				e := conn.SetReadDeadline(deadline)
+				if e != nil {
+					logger.Infof("could not set read deadline on command request: %v", e)
+					return
+				}
+				for {
+					if time.Now().After(deadline) {
+						conn.Write(marshalOrInternalError(TimeoutError))
+						return
+					}
+					rune, _, err := r.ReadRune()
+					if err != nil {
+						logger.Debugf("connection read error: %v", err)
+						if errors.Is(err, os.ErrDeadlineExceeded) {
+							conn.Write(marshalOrInternalError(TimeoutError))
+						} else {
+							conn.Write(marshalOrInternalError(ConnError))
+						}
+						return
+					}
+					b = append(b, byte(rune))
+					switch rune {
+					case '{':
+						depth++
+					case '}':
+						depth--
+					}
+					// Must check here because the first pass always depth = 0
+					if depth == 0 {
+						break
+					}
+				}
+				var req Request
+				err := json.Unmarshal(b, &req)
+				if err != nil {
+					conn.Write(marshalOrInternalError(BadRequestError))
+					return
+				}
+				handlersMu.RLock()
+				defer handlersMu.RUnlock()
+				handler, ok := handlers[req.Command]
+				if !ok {
+					conn.Write(marshalOrInternalError(CmdNotFoundError))
+					return
+				}
+				resp, err := handler(b)
+				if err != nil {
+					resp = marshalOrInternalError(Response{Status: HandlerError.Status, StatusMessage: err.Error()})
+				}
+				conn.Write(resp)
+			}(conn)
+		}
+	}()
+	c.srv = srv
+	return nil
+}
+
+// Wait does not start the Server, it waits for a signal to begin listening
+// on the given context, as long as it is valid. To signal the server to start
+// or stop listening, call Server.Start() or Server.Stop(). To signal the
+// Server to stop waiting, call Server.Close(). This function is
+// thread-safe, but calling it multiple times is generally a bad idea, as Close()
+// will only close a single Wait() call.
+func (c *Server) Wait(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
```

Comment:
we are gracefully returning, just log it and return nil. The caller is logging the error so only return an error when it's actually an error.

Reply:
Done.